frontend/src/metabase/containers/AdHocQuestionLoader.jsx

```diff
@@ -0,0 +1,124 @@
+/* @flow */
+import React from "react";
+import { connect } from "react-redux";
+
+// things that will eventually load the quetsion
+import { deserializeCardFromUrl } from "metabase/lib/card";
+import { loadMetadataForCard } from "metabase/query_builder/actions";
+import { getMetadata } from "metabase/selectors/metadata";
+
+import Question from "metabase-lib/lib/Question";
+
+// type annotations
+import type Metadata from "metabase-lib/lib/metadata/Metadata";
+
+/*
+ * AdHocQuestionLoader
+ *
+ * Load a transient quetsion via its encoded URL and return it to the calling
+ * component
+ *
+ * @example
+ *
+ * Render prop style
+ * import AdHocQuestionLoader from 'metabase/containers/AdHocQuestionLoader'
+ *
+ * // assuming
+ * class ExampleAdHocQuestionFeature extends React.Component {
+ *    render () {
+ *      return (
+ *        <AdHocQuestionLoader questionId={this.props.params.questionId}>
+ *        { (question) => {
+ *
+ *        }}
+ *        </SavedQuestion>
+ *      )
+ *    }
+ * }
+ *
+ * @example
+ *
+ * The raw un-connected component is also exported so we can unit test it
+ * without the redux store.
+ */
+
+type Props = {
+  children: Function,
+  loadMetadataForCard: Function,
+  metadata: Metadata,
+  questionHash: string,
+};
+
+type State = {
+  // the question should be of type Question if it is set
+  question: ?Question,
+};
+
+export class AdHocQuestionLoader extends React.Component {
+  props: Props;
+
+  state: State = {
+    // this will store the loaded question
+    question: null,
+  };
+
+  componentWillMount() {
+    // load the specified question when the component mounts
+    this._loadQuestion(this.props.questionHash);
+  }
+
+  componentWillReceiveProps(nextProps: Props) {
+    // if the questionHash changes (this will most likely be the result of a
+    // url change) then we need to load this new question
+    if (nextProps.questionHash !== this.props.questionHash) {
+      this._loadQuestion(nextProps.questionHash);
+    }
+  }
+
+  /*
+   * Load an AdHoc question and any required metadata
+   *
+   * 1. Decode the question via the URL
+   * 2. Load any required metadata into the redux store
+   * 3. Create a new Question object to return to metabase-lib methods can
+   *    be used
+   * 4. Set the component state to the new Question
+   */
+  async _loadQuestion(questionHash: string) {
+    const card = deserializeCardFromUrl(questionHash);
+    // pass the decoded card to load any necessary metadata
+    // (tables, source db, segments, etc) into
+    // the redux store, the resulting metadata will be avaliable as metadata on the
+    // component props once it's avaliable
+    await this.props.loadMetadataForCard(card);
+
+    // instantiate a new question object using the metadata and saved question
+    // so we can use metabase-lib methods to retrieve information and modify
+    // the question
+    const question = new Question(this.props.metadata, card);
```

Comment:
One thing I'm realizing now is that if the `metadata` is later updated after `question` is added to the state then it won't update with the new metadata. This probably won't happen often but we should handle it.

There are a few ways

1. Store both `question` and `card` in `state`, add the following to `componentWillReceiveProps`:
```
if (nextProps.metadata !== this.props.metadata && this.state.card) {
  this.setState({ question: this.state.card && new Question(nextProps.metadata, this.state.card) })
}
```

2. Only store `card` in `state`, create a reselect selector for `question` in the constructor:
```
this._getQuestion = createSelector(
  [(s, p) => s.card, (s, p) => p.metadata],
  (card, metadata) => card && new Question(metadata, card)
)
```
then in `render` do
```
return this.props.children(this._getQuestion(this.state, this.props));
```

3. Implement `shouldComponentUpdate` (or possibly just use `React.PureComponent` but see this caveat https://reactjs.org/docs/render-props.html#be-careful-when-using-render-props-with-reactpurecomponent) to check if `card` or `metadata` has changed:
```
return nextProps.metadata !== this.props.metadata || nextState.card !== this.state.card
```
and in `render` just do:
```
return this.props.children(this.state.card && new Question(this.props.metadata, this.state.card));
```

Reply:
Do you have a preference for which approach we should take? It seems like 1 might be the easiest to grok.